File: algo/31. Next Permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    if (nums.length==0) {
        return [];
    }
    
    var currIndex = nums.length - 1;
    while (currIndex > 0) {      
        if (nums[currIndex] > nums[currIndex-1]) {
            break;
        }
        currIndex--;
    } 
    
    
    if (currIndex > 0) {
        var minIndex = currIndex;
        var currIndex2 = currIndex;
        while (currIndex2 < nums.length) {
            if (nums[currIndex2] > nums[currIndex-1]) {
                min = nums[currIndex2];
                minIndex = currIndex2;
            } else {
                break;
            }            
            currIndex2++;
        }
        // swap the numbers
        var temp = nums[minIndex];
        nums[minIndex] = nums[currIndex-1];
        nums[currIndex-1] = temp;
    }
    
    // swap the numbers in the tail
    var left = currIndex;
    var right = nums.length - 1;
    while (left<right) {
        temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
        left++;
        right--;
    }
};

File: round3/31. Next Permutation.cs
public class Solution {
    public void NextPermutation(int[] nums) {
        for (int i = nums.Length - 2; i >= 0; i --) {
            if (nums[i] < nums[i+1]) {
                for (int j = nums.Length - 1; j > i; j --) {
                    if (nums[j] > nums[i]) {
                        int temp = nums[j];
                        nums[j] = nums[i];
                        nums[i] = temp;                        
                        
                        Array.Reverse(nums, i + 1, nums.Length - i - 1);
                        return;
                    }
                }                
            }
        }
        
        Array.Reverse(nums);
    }
}